src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/utils/CardinalityEstimation.java

```diff
@@ -0,0 +1,265 @@
+
+        ppCardinalityTable.add( pp, cardinality );
+        return cardinality;
+    }
+
+    public int getJoinCardinalityEstimation( final PhysicalPlan pp, final FederationAccessManager mgr ) throws FederationAccessException {
+        if ( ppCardinalityTable.contains(pp) ){ return ppCardinalityTable.get(pp); }
+
+        final PhysicalOperatorForLogicalOperator lop = (PhysicalOperatorForLogicalOperator) pp.getRootOperator();
+        if ( !(lop.getLogicalOperator() instanceof LogicalOpJoin)){
+            throw new IllegalArgumentException();
+        }
+
+        final PhysicalPlan pp1 = pp.getSubPlan(0);
+        final PhysicalPlan pp2 = pp.getSubPlan(1);
+
+        final int cardinality = joinCardinality( pp1, pp2, mgr );
+        ppCardinalityTable.add(pp, cardinality);
+
+        return cardinality;
+    }
+
+    public int getTPAddCardinalityEstimation( final PhysicalPlan pp, final FederationAccessManager mgr ) throws FederationAccessException {
+        if (ppCardinalityTable.contains(pp)){ return ppCardinalityTable.get(pp); }
+
+        final PhysicalOperatorForLogicalOperator pop = (PhysicalOperatorForLogicalOperator) pp.getRootOperator();
+        final LogicalOperator lop = pop.getLogicalOperator();
+        if ( !(lop instanceof LogicalOpTPAdd)){
+            throw new IllegalArgumentException();
+        }
+
+        final PhysicalPlan pp1 = pp.getSubPlan(0);
+        final PhysicalPlan reqTP = formRequestBasedOnTPofTPAdd( (LogicalOpTPAdd) lop );
+
+        final int cardinality = joinCardinality( pp1, reqTP, mgr );
+        ppCardinalityTable.add( pp, cardinality );
+
+        return cardinality;
+    }
+
+    public int getBGPAddCardinalityEstimation( final PhysicalPlan pp, final FederationAccessManager mgr ) throws FederationAccessException {
+        if ( ppCardinalityTable.contains(pp) ){ return ppCardinalityTable.get(pp); }
+
+        final PhysicalOperatorForLogicalOperator pop = (PhysicalOperatorForLogicalOperator) pp.getRootOperator();
+        final LogicalOperator lop = pop.getLogicalOperator();
+        if ( !(lop instanceof LogicalOpBGPAdd)){
+            throw new IllegalArgumentException();
+        }
+
+        final PhysicalPlan pp1 = pp.getSubPlan(0);
+        final PhysicalPlan reqBGP = formRequestBasedOnBGPofBGPAdd( (LogicalOpBGPAdd) lop );
+
+        final int cardinality = joinCardinality( pp1, reqBGP, mgr );
+        ppCardinalityTable.add( pp, cardinality );
+
+        return cardinality;
+    }
+
+    private int joinCardinality( final PhysicalPlan pp1, final PhysicalPlan pp2, final FederationAccessManager mgr ) throws FederationAccessException {
+        final Set<Var> certainJoinVars = ExpectedVariablesUtils.intersectionOfCertainVariables( pp1.getExpectedVariables(), pp2.getExpectedVariables() );
+        final Set<Var> possibleJoinVars = ExpectedVariablesUtils.intersectionOfPossibleVariables( pp1.getExpectedVariables(), pp2.getExpectedVariables() );
```

Comment:
The possible join variables is the union of **all** variables in both plans, minus the certain join variables.